### ISSUE TITLE
[@typescript-eslint/restrict-plus-operands] false positive when using conditional/ternary operator

### ISSUE THREAD

**Issue by KilianKilmister** [typescript-eslint/typescript-eslint#2547]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

rule:

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": ["error", { "checkCompoundAssignments": true }]
  }
}
```

example function

```TS
function someFunc (someBoolean: boolean) {
  return 'string a and ' + someBoolean ? 'string b' : 'string c'
}
```

`compilerOptions` in `tsconfig`

```json5
{
  "lib": [
    "esnext"
  ],
  "alwaysStrict": true,
  "stripInternal": true,
  "assumeChangesOnlyAffectDirectDependencies": true,
  "declaration": true,
  "declarationDir": "@types",
  "declarationMap": true,
  "diagnostics": true,
  "forceConsistentCasingInFileNames": false,
  "importsNotUsedAsValues": "error",
  "incremental": true,
  "module": "esnext",
  "moduleResolution": "node",
  "outDir": "dist",
  "plugins": [],
  "pretty": true,
  "rootDir": ".",
  "baseUrl": ".",
  "sourceMap": true,
  "target": "ESNext",
  "useDefineForClassFields": true,
  "experimentalDecorators": true,
  "strict": true,
  "strictBindCallApply": true,
  "strictFunctionTypes": true,
  "strictNullChecks": true,
  "noImplicitAny": true,
  "strictPropertyInitialization": true,
  "noImplicitReturns": false,
  "noImplicitThis": true,
  "allowSyntheticDefaultImports": true,
  "emitDecoratorMetadata": true,
  "composite": true,
  "preserveSymlinks": true
},
```

**Expected Result**

no error

**Actual Result**

> error    Operands of '+' operation must either be both strings or both numbers. Consider using a template literal                                                                                                                                                       [@typescript-eslint/restrict-plus-operands](https://github.com/typescript-eslint/typescript-eslint/blob/v4.0.1/packages/eslint-plugin/docs/rules/restrict-plus-operands.md)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `3.10.1` |
| `TypeScript`                       | `4.1.0-dev.20200911` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.10` |
